IOfile refactor

mkosunen opened the pull request:
Should address issues: 
#16
#18  
Ping @sporrasm @vlahtin

mkosunen committed:
Start refactoring

mkosunen committed:
Fix _bin2int

Ordering was incorrectly reversed

mkosunen committed:
Actually use self.big_endian

Commit:
Fix crashes when only sample type outputs are defined

- Crashes were due to missing trigger signal that was not read to
  self.iofile_eventdict, since trigger signal was not read unless
  explicitly defined to be an event type output

- On the other hand, explicitly defining triggers as event outputs
 caueses columns to overflow in the print file (i.e. num_cols is
 calculated incorrectly) in some cases

- The most convienient solution is to just read the output file once
BEFORE reading the individual IOs

sporrasm commented:
Fixed #22 by refactoring print-file reading away from the spice_iofile, since it is not related to any single IO, but all of them. This may break other things, I guess.

Also, fix #16

vlahtin committed:
Add support for Spectre X.

sporrasm committed:
Merge pull request #24 from TheSystemDevelopmentKit/spectrex_support

Add support for Spectre X.

Commit:
Add option to use psf for file reading

vlahtin commented:
Tried AC; does not work yet. The reason is that on line 180, the io type is read and it uses tran_analysis_name always for event based results (even though AC can also use those). 

Need to implement some method to extract whether or not the simulation was actually AC or Transient. This information is atleast given to spice_simcmd, so have to figure out some method to get it here as well. Did not have time today to debug this further, but leaving this as documentaiton if someone else has the time to figure this out before me.

sporrasm commented:
I see. The way I see it is that we three options

1) Either use the same analysis name for all simulations, which is probably not a good idea
2) Infer output file name from properties of spice_iofile. For example, I think AC simulations should have iotype='event' and datatype='complex'
3) Detect which simulation type we are running and infer output file name based on that, which is the same what @vlahtin proposed above.

To me option 1 is the worst and option 2 is not very explicit and hence might be difficult to debug. It probably makes sense to detect the simulation type and infer analysis name from there directly. This will mean that running two different simulations on the same netlist will not work, but I don't think that has been supported for a long time anyway.

vlahtin committed:
Check which simutype it is and name the file correctly

vlahtin commented:
https://github.com/TheSystemDevelopmentKit/spice/pull/17/commits/da34a259d79cd42b587b11adaa45934e602471e0 should fix the aforementioned problem.

Commit:
use libpsf to read transient

Commit:
Merge

sporrasm commented:
Using libpsf to read in transient and AC. Tested with transient. This now has the downside that sparam, etc, still use psf_utils, which we should probably get rid of, if libpsf is superior to psf_utils. See #25 

@vlahtin can you test and see if it is faster now? Please remember to pip3 install --user libpsf

Commit:
Fix AC

sporrasm commented:
Tested also AC with a nonsense example. At least it doesn't crash.

We should probably aim to make release tests for all analysis types to aid with the testing.

Commit:
Output file reading for event, sample and time outputs is now simulator specific

Commit:
Allow reading from multiple files

Commit:
Multiple output files supported, at least in theory

Commit:
Move strobing related stuff under spectre

Commit:
iofile reading is now done inside iofile, as it should be

Commit:
Fixup

Commit:
Extract powers only in transient

MiikkaMaki committed:
fixing pac and pss

MiikkaMaki committed:
Merge branch 'iofile-refactor' of github.com:TheSystemDevelopmentKit/spice into iofile-refactor

Commit:
iofile read operation is now simulator agnostic

- This required to get rid of iotypes 'psfascii_pss' and 'psfascii_pac',
which are simulator specific. They are now handled as events. The
reason is that AC simulation output (freq, amplitude pairs) is also an event.
PSS and PAC output similar type of data. The frequency is fundemntal or
harmonic in PSS and in PAC it is fund +/- nharmonics * pss_fund.

- Seems to break noise simulations...

Commit:
Remove breakpoint

Commit:
Fixup noise simulation

sporrasm commented:
Okay, tested this together with @MiikkaMaki. transient, dc, ac, sp, noise, pss and pac simulations seem all to work. Of course, further testing would always be nice to catch possible uncaught bugs (mention @vlahtin if you have the time).

iofile reading operation should now be simulator agnostic once again, which is good news.

vlahtin committed:
Fix spectre X preset names

Commit:
Fixup?

vlahtin committed:
Add option to use psf with psf utils and libpsf for transient. For other analyses use psf utils

vlahtin committed:
Merge branch 'iofile-refactor' of github.com:TheSystemDevelopmentKit/spice into iofile-refactor

vlahtin commented:
Added the option to select either binary parsing (using libpsf) or ascii parsing (using psf_utils) when use_psf=True. Also made it so, that psf_utils is used for all other analyses (when use_psf is True), as the output file size is not expected to grow as significantly as transient, thus yielding reliable results with reasonable speed.

This change came as libpsf has problems with memory leakage with large output file sizes, and unavailable in some linux distributions. The psf_utils is however quite slow, so if large file sizes in transient are expected, either libpsf or print file parsing may be the best option, which should now be available all according to user selection

@MiikkaMaki can you test that your SP, noise, PSS, PAC testbench still works?

vlahtin committed:
Force print file generation if not use_psf

Commit:
Use black to format code

vlahtin committed:
Fix bug which crashes if rawfmt is not in spiceoptions

vlahtin committed:
Noise must have psfbin type rawfmt

vlahtin committed:
Fix PAC

vlahtin committed:
Fix PSS data parsing

vlahtin committed:
Explicitly import from abc

vlahtin committed:
Dont add the brackets if the sample IO is for one IO instead of a bus

vlahtin commented:
Hi,
happened to do something that has not been previously done with Spice. Lets add the fix to that to this merge as well: 
https://github.com/TheSystemDevelopmentKit/spice/pull/29

Please test that as well before merge. It works in my special case, but if someone has another case with sample IOs, please try that.

mkosunen commented:
Notes for the release:As the lint-based code reformat policy has been added to v1.14_RC already, I suggest that you do octopus merge (and some tests) by merging v1.14_RC to this branch and re-test before merging this to v1.14_RC, if not done already.

mkosunen commented:
Note for the future developments: If there is a need for a iofiletype where the first column is frequency, (for example in AC-analysis) then the most consistent way to handle this is to add io type 'freq' with an assumption that the data in the other columns are complex. All the other formats address the problem how the time in the first columns is interpreted.

Opened #31 for this, does not need to be addressed now.

Also: AC response is not a 'signal' as in sense of conveying data from a block to another and thus not strictly an IO. To me it seems that it should be handled with some kind of 'extract' rather than IO's. Freq type signal would be for example an OFDM frame.

mkosunen commented:
Further notes fro the release: To reduce the number of conflicts at merge, regrdless of which way it is done first, you may run 
`black -l 80  --check --diff --color <list of python files> `
to check the syntax differences and then correct them with 
`black -l 80  <list of python files>`

MiikkaMaki committed:
fix interactive with no other flags

vlahtin committed:
Merge pull request #29 from TheSystemDevelopmentKit/sample_ioname_fix

Dont add the brackets if the sample IO is for one IO instead of a bus

vlahtin committed:
Merge branch 'v1.14_RC' into iofile-refactor
